### ISSUE TITLE
Missing day returns last day of the month, prefer 0.

### ISSUE THREAD

**Issue by jwaynes** [akoumjian/datefinder#36]
Using Python 3.5.2 and datefinder 0.6.0.
textstring = "Oct , 1881"
find_dates returns '1881-10-31 00:00:00'
I would have preferred '1881-10-00 00:00:00'

Is there a way to force the default day?

**Comment by akoumjian** [comment]
My guess, without testing, is that this is simply the default behavior for dateutil. It's worth noting that the preferred date is not a valid datetime (day 00 is not in the calendar). I couldn't say why dateutil returns the last day rather than the first day of the month (01). 

It's worth noting that you can now pass a ``base_date`` parameter to datefinder, and so with a missing date this may fill in the gaps accordingly (again, I haven't tested). https://github.com/akoumjian/datefinder/blob/master/datefinder.py#L270

**Comment by jwaynes** [comment]
I've chosen to do a find_dates with strict=True, followed by a find_dates with strict=False.
If it fails on the first but passes the second, I can assume it is missing the day and can then handle the case as needed.

**Comment by pgrenon** [comment]
There is still an issue around missing days or more generally perhaps missing parts of dates. In the case of missing days, the workaround does not work as it does not allow distinguishing between missing day and missing year.

The behaviour seems to be to replace with the components of the current date.

## guesses missing day and replaces with today's day
list(datefinder.find_dates('Oct 1883', strict=False))
[datetime.datetime(1883, 10, 25, 0, 0)]

## guesses missing year and replaces with today's year
list(datefinder.find_dates('Oct 18', strict=False))
[datetime.datetime(2019, 10, 18, 0, 0)]

Is there are concept of date format that can be applied to find_dates in order to guide it when there is a notion of what is expected?
E.g.,
list(datefinder.find_dates('Oct 18', format="%M%YY"))

**Comment by stevendavis** [comment]
This is an old issue, but I'm posting my solution in case it helps someone.  Here is my solution to the missing day and some other issues as well:


```
year_regex = re.compile(r"(19|20)\d\d")

def parse_date_str(line):
    """Parse one or more dates from a line of text.  Return the first date found.

    Parameters
    ----------
    line : str
        String possibly containing a date.

    Returns
    -------
    dt : datetime.date
        The first date found, if any, otherwise None is returned.
    """
    # Don't let datefinder find dates when the year is missing or invalid, for example "1 1 1".
    # This test also improves parsing speed by rejecting the unwanted lines quickly.
    match_object = year_regex.search(line)
    if not match_object:
        return None

    # The datefinder package fails to find dates when preceeded by a tag with colon, for example "date: 1/1/2000".
    # This unconditional replace may be a bit too heavy handed, keep an eye on it.
    line = line.replace(': ', ' ')

    # We need to parse dates in a non-strict way because strict disallows leading zeros, for example, 01/01/2020.
    # However, when strict=False, a missing date component is replaced by today's date component.
    # We don't want that behavior, so we parse twice with two different filler-dates and make sure
    # the results are the same.
    today = datetime.datetime.today()
    datetimes = list(find_dates(line, strict=False, base_date=today))
    if len(datetimes) == 0:
        return None

    other_date = today - datetime.timedelta(days=365+31+15)  # must change the year and month and day
    datetimes2 = list(find_dates(line, strict=False, base_date=other_date))

    for dt1, dt2 in zip(datetimes, datetimes2):
        if dt1 == dt2:
            return dt1.date()
        else:
            # Not sure what is the best behavior here, another possible behavior is to return None
            # and stop looking for other dates in this line of text.
            pass

    return None

```

**Comment by akoumjian** [comment]
This is working as intended. A couple items to consider for others finding this page:

1. You can pass in a `base_date` to `datefinder.find_dates`. Any missing component from a date will be filled in with the value from that `base_date`.  
2. This library returns `datetime` objects and there's no notion of masking missing values with those. Perhaps in some future release we would consider returning some kind of proxy object that allows for masking of missing values, but for now you get datetimes and you have to have a year, month, day of the month, etc. If you want year-month combinations to return the first of the month, make sure your base date is the first of any month.

Separately, @stevendavis found this one:
`# The datefinder package fails to find dates when preceeded by a tag with colon, for example "date: 1/1/2000".`

That is _not_ intended behavior, which is not fixed (#172 )
